Deployments

cavemanpi opened the pull request.

cavemanpi committed:
Improves documentation clarity. Fixes copyright notice.

cavemanpi committed:
Adds perl 5.22 and 5.24 to travis tests.

cavemanpi committed:
Change homepage and copyright info in dist.ini

cavemanpi committed:
Consolidate some duplicated code.

cavemanpi committed:
Works around a bug in kuberntes' namespace self links.

cavemanpi committed:
Add bare bones class setup for Deployments.

cavemanpi committed:
White space.

cavemanpi committed:
Kubernetes api versioning looks to be on a per object level. Let the user specify.

cavemanpi committed:
Add capabilities to create and list deployment resources.

cavemanpi committed:
Delegates listing of namespace scoped resources to the default namespace.

cavemanpi committed:
Adds support for label selector expressions.

cavemanpi committed:
Start adding awareness of server version.

cavemanpi committed:
What do we do when we want a method two places, but inheritance doesn't make sense? Role with it.

cavemanpi committed:
Adds knowledge of namespaces to resources.

cavemanpi committed:
Adds a resource catalog so we know where resources actually live in the API by server version.

cavemanpi committed:
Adds tests for Deployments

cavemanpi committed:
Adds a method to scale a deployment.

cavemanpi committed:
Whitespace

cavemanpi committed:
Adds knowledge of replicasets to the module.

cavemanpi committed:
Adds tests surrounding replica sets.

cavemanpi committed:
Update Changes. Bump version. Let dzil rebuild the README.

cavemanpi committed:
Documentation improvements

cavemanpi committed:
Adds get methods for deployments and replica sets. Improves tests.

cavemanpi committed:
Delegate get_deployment and get_rs to default namespace.

cavemanpi committed:
Let dzil regenerate the README

jimbobhickville commented on lib/Net/Kubernetes/Resource/Deployment.pm:
```diff
@@ -0,0 +1,27 @@
+with 'Net::Kubernetes::Resource::Role::HasPods';
+with 'Net::Kubernetes::Resource::Role::HasReplicaSets';
+
+sub scale {
```
Missing POD

jimbobhickville commented on lib/Net/Kubernetes/Resource/Deployment.pm:
```diff
@@ -0,0 +1,27 @@
+
+sub scale {
+    my($self, $replicas, $timeout) = @_;
+    $timeout ||= 5;
```
Can we make the default timeout configurable somehow?

jimbobhickville replied:
Also should probably be `//=` because 0. Should 0 be "wait forever" and -1 "return immediately" or vice-versa? I feel like having both options is useful, but I'm less certain on how the caller should indicate that.

cavemanpi replied:
Likely the best way to do that is to put an attribute in the APIAccess role and advertise it for the constructor of Net::Kubernetes. 

I like 0 for return immediately and -1 for wait forever.

jimbobhickville commented on lib/Net/Kubernetes/Resource/Deployment.pm:
```diff
@@ -0,0 +1,27 @@
+    while((time - $st) < $timeout){
+        my $pods = $self->get_pods;
+        if(scalar(@$pods) == $replicas){
+            return "scaled";
```
I find the return signature a bit weird. It's a string here, but boolean ints otherwise.

cavemanpi replied:
Yeah, I agree. I was trying to keep this consistent with the scale method in replication controllers. In fact, it's really the same method. Could extract that out to yet another role.

jimbobhickville commented on lib/Net/Kubernetes/Resource/Role/HasReplicaSets.pm:
```diff
@@ -0,0 +1,47 @@
+	if ($res->is_success) {
+		my $set_list = $self->json->decode($res->content);
+		my @sets;
+		warn Dumper($set_list);
```
Leftover debug?

cavemanpi replied:
Yep, thought I got them all. Fixing shortly.

jimbobhickville commented on lib/Net/Kubernetes/Resource/Role/HasReplicaSets.pm:
```diff
@@ -0,0 +1,47 @@
+			$create_args{password}    = $self->password if($self->password);
+			$create_args{url}         = $self->url;
+			$create_args{base_path}   = $rs->{metadata}{selfLink};
+			$create_args{ssl_cert_file} = $self->ssl_cert_file if($self->ssl_cert_file);
```
Your lining up stops lining up.

jimbobhickville commented on lib/Net/Kubernetes/Role/APIAccess.pm:
```diff
@@ -121,30 +130,58 @@ around BUILDARGS => sub {
+sub _build_server_version {
+	my $self = shift;
+
+	my($version_info) = $self->send_request($self->create_request(GET => $self->url . '/version'));
```
Does this get shared across multiple objects from the same client session somehow? Seems like it should.

cavemanpi replied:
Sort of, not really. One structural problem this module has is the fact that most objects are also api clients. That leads to having to pass on all of the api related attributes upon creation of another object. When the server_version is passed on, an object will not ask the server for its version again.

I made sure to pass that variable on in the ResourceFactory role, but I see that I've missed it in the HasPods and HasReplicaSets roles.

If I ever get around to it, I think a major point release would be in order to restructure this API model, making the client an actual object and passing the client around.

jimbobhickville commented on lib/Net/Kubernetes/Role/ResourceCatalog.pm:
```diff
@@ -0,0 +1,76 @@
+requires 'url';
+
+my %endpoint_catalog = (
+	1.5 => {
```
Do we not care about earlier versions any more or does the default behavior work for any prior version?

cavemanpi replied:
I've tested against 1.2 and 1.5. They both seem to work for resources I'd expect to see in each version. The default behavior is what's expected as of 1.2. The problem is that I can't find documentation for the intermediate versions. 

The following documentation only starts at 1.5:
https://kubernetes.io/docs/reference/

jimbobhickville commented on lib/Net/Kubernetes/Role/ResourceCatalog.pm:
```diff
@@ -0,0 +1,76 @@
+	my $server_version = $self->server_version;
+
+	$resource          = lc($resource);
+	my $resource_api   = $endpoint_catalog{$server_version}{$resource} || 'api/v1';
```
I feel like we should find the previous version if a version is higher than the highest we know about, because that is likely to be the closest match.  So server 1.7 would get the 1.6 urls until you can update it with any changes for 1.7. Either that or error with unsupported version or something if the server version is higher than you expect.

cavemanpi replied:
Good idea!

cavemanpi replied:
This should do that: 347700f711dd3be67887286bae7da7ed6e1fe2aa

jimbobhickville reviewed (commented):
Nothing too major, just some nits and questions.

cavemanpi committed:
Remove warn with Dumper.

cavemanpi committed:
White space

cavemanpi committed:
Send server version along to pods and replica sets.

cavemanpi committed:
If the server version is higher than the catalog knows, use the most recently known version.

cavemanpi committed:
Improve tests.

cavemanpi committed:
Make scale timeout control more flexible.

cavemanpi committed:
Fix broken pod

cavemanpi committed:
v1.04

  - Adds support for deployments and replica sets.

  - Begin deprecation of api_version as various resources have different version info.

  - Adds server version detection.

  - Adds internal cataloging of resource endpoints to server versions.

  - Net::Kuberentes now delegates all namespaced resource listing to the default namespace.

  - Consolidate some repeated code.

  - Add support for match expressions in label selectors.
